Capistrano deploy: add task to run 'one-time' tasks

weedySeaDragon opened the pull request:
### PT Story:  Capistrano deployment: automatically check for and log 1 time tasks to run 
https://www.pivotaltracker.com/story/show/166136006

I refactored.  I wrote this so that it could eventually be a _gem._  (The main work that needs to be done is to separate the _logging_ -- right now it depends on the `ActivityLogger.`  This dependency needs to be addressed in order for this to be made into a gem.)

**Here are the main classes and an overview on how they interact:**

![OneTimeTasker classes UML - OneTimeTasker](https://user-images.githubusercontent.com/673794/60362464-d25d0880-9995-11e9-9b9a-2a871c4e005c.png)


### Changes proposed in this pull request:
1. Added a new task that is called during capistrano deploy that will invoke "one time" tasks: `shf:deploy:run_onetime_tasks`
   When a "one-time" task is invoked, the success or failure is logged.
   If the task _fails_ it will be invoked again the next time deployment is done.  It will be invoked until it has been run successfully.  (see `OneTimeTasksFinder` below)

1. `OneTimeTasksFinder` class finds tasks that should be run:
   - the `.rake` file for the task is in a directory correctly named for this year and this quarter (YYYY-Qn)
   - the task has _not_ already been run successfully.  
      A task has already been successfully run if there is an entry in the log that matches `"One-time task #{task_name} was run on"`

   Will get _all tasks_ in a file (can have multiple tasks in a file).

2. Sequence during deployment: This is called _after_ the `load_conditions` task.  Here's the sequence:
   1. migrate
   1. run_load_conditions
   2. **run_one_time_tasks**
      - set existing one-time tasks to 'already ran' only if this hasn't _already_ been done
   3. create the release directory, remove test files, restart.

2. clarified and DRYed  added comments to `deploy.rb`
3. `ActivityLogger` now always returns the log, even when given a block


### Ready for review:

weedySeaDragon committed:
class to find and gather tasks to run one time now

weedySeaDragon committed:
task to gather and invoke all one time tasks to run now

weedySeaDragon committed:
Merge branch 'av-develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
Merge branch 'av-develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
ensure tasks are reenabled so they can be invoked; add comment

weedySeaDragon committed:
update to use 'logfilepath' for ActivityLogger

weedySeaDragon committed:
create a task given the body of the task (refactor); set logfile name

weedySeaDragon committed:
if a task fails, it will be (re-)invoked the next time, til it succeeds

weedySeaDragon committed:
Merge branch 'av-develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
ActivityLogger always return the log (even when a block is given)

weedySeaDragon committed:
rename; fix spec (stub so no rake tasks are found)

weedySeaDragon committed:
add run one time tasks; comment and re-org; dry (invoking tasks)

weedySeaDragon committed:
spec to show it handles multiple tasks in 1 file; comments

patmbolger commented:
The architecture looks great.  My one concern has to do with preventing any rake task from being run more than once.  We try to make all rake tasks idempotent but we can't guarantee that every time - especially as new data is added to the database (thus an old rake task could be run against - and alter - that new data).

Right now this relies on checking the log file.  That's not good enough because the `log` directory is not maintained in git.  

Some thoughts on how to address this:

1. Put the log file in the `tasks` directory, to ensure it is maintained in git, or,
2. Store the run event in a DB table.

Also, I'd recommend changing the file name for any task that is run from `<task_name>.rake` to `<task_name>.rake.run`.  That way it will never be picked up again given the logic here.

patmbolger reviewed (changes requested):
Please see comment.

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon commented:
Excellent ideas as usual @patmbolger  :-)

Misc:
- yes, using the log file is a pretty cheesy way to see if a task has been run.  
- We should never use git to store log files:  There is personally identifiable information in there -- which would then become available on our public Github repo. (And it would massively violate GDPR.)
   - as an aside: We do back up the log files, so we do have a snapshot of the state of the app in that sense

- Cannot mark a task as being run at the file level (e.g. renaming a file from `.rake` to `.rake.run`) because you can have multiple tasks in a .rake file, and some in the file might run successfully and some might not.   Have to track success/failure of a `Rake::Task` at the `Rake::Task` level

### Record (persist) whether or not a Task was successful:
I agree, it'd be much better if we actually persist the info about whether a task ran successfully in the db.  We're using the database as our persistent store.  We should keep all data that needs to be persistent in that.  We shouldn't split it up and spread the responsibility for persistence it out among miscellaneous files _and_ the database unless we have an incredibly compelling reason. 


So I propose :

**Create a model to represent tasks that we try to run** (maybe named `TaskRunAttempt` ?)
	• name of the task
	• when did we attempt to run it?
	• was the attempt successful?
	• source of the task (e.g. rake file name)

- We have 2 Use cases for this:
1. We need to know whether or not a task needs to be run, either for the first time or for a repeated attempt.
2. Provide us with information so we understand what tasks were run and when, as a part of understanding how the application data may have been manipulated historically.

_Other use cases?  Will the 4 attributes I proposed above for the model provide us with the info we need for those use cases?_

This model becomes part of the infrastructure for the App.  It's not part of the domain, but is required for maintaining and using the app. 
- There is a gem that pretty much does this: https://github.com/theSteveMitchell/after_party.  If we use an ActiveRecord to track success/failure, I think we're doing a better job of it.

(ETA: directory name year/quarter):

Given the above, what is the value provided by the task directory naming? e.g. `/YYYYQ1` etc
- it is helpful in organizing task files
- is it really needed for any information about whether or not a task should be run?  That is:  either a task has already been run successfully or not (in which case we need to try again).
   - plus I realized that relying on this  naming convention actually would fall apart if a task failed on the last day of a quarter; something I didn't cover in this PR 

So we can continue to put subdirectories under the main `/one-time` task directory and name them whatever we want, but directory names are not going to play any part in determining whether or not a task should be run.


Thoughts?

weedySeaDragon committed:
move to OneTimeTaskRunner:: (rename files, classes); + config initializer

patmbolger commented:
@weedySeaDragon Thinking out loud ...

1. My thought on using the log file was based on it being specific to (and only to) these tasks (via the naming action `LogfileNamer.name_for(SHFDEPLOY_LOG_NAME)`).  So, no user-specific data and no GDPR pitfalls. Also, since the log file was to exist in the `tasks/one_time` directory, it would be maintained in git and be available to any developer who cloned the project.

2. Using the DB is pretty good for production (we'd need a one-time task to add the current baseline data - a bit ironic).  My misgiving is that we're depending _solely_ on that table to have valid data, in order to prevent possibly very dangerous tasks from being run automatically, so ....

3. That's why I think we need another, "suspenders" mechanism, such as renaming rake task filenames to prevent them from ever being run again.   For that to work, we could impose the requirement that each `one_time` task file have only _one_ task (we could enforce this in your code).  This also would prevent the naive new developer from running `shf:deploy:run_onetime_tasks` locally and getting into confusing trouble. 

4. Re the directory naming convention - your 2nd use case - "Provide us with information so we understand what tasks were run and when, as a part of understanding how the application data may have been manipulated historically." is germane here.  The DB records can tell us when a task was run.  However, I think it's also very useful to view the directories and get a rough idea as to how the data has evolved as the app has grown and we've restructured models and associations.  So, from my POV there is value in that mechanism and no gain in dropping it.  (and yes, if a rake task fails in the last day of that quarter then 1) we move the task to the next quarter's directory, or, 2) we end up running that task the following week - but given the macro view the directory structure gives us, that really doesn't matter.

weedySeaDragon committed:
OneTimeTasker::TaskAttempt model (namespaced under OneTimeTasker)

weedySeaDragon committed:
TaskAttempt: add 'notes' attribute

weedySeaDragon committed:
rename to 'simple rake task files maker' to better reflect intent

weedySeaDragon committed:
specify table name; was_successful validation; readability methods

weedySeaDragon committed:
rename initializer; is for the OneTimeTasksFinder

weedySeaDragon committed:
rename module to "OneTimeTasker" (= directories change names)

weedySeaDragon committed:
Merge branch 'av-develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
add yarn.lock (even tho it's empty)

weedySeaDragon committed:
rename to just 'TasksFinder' (because the module name has 'OneTime' already

weedySeaDragon committed:
Merge branch 'av-develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
handle classes with module prefix (replace '::' with '_')

weedySeaDragon committed:
initializer should be for TasksRunner (rename, etc)

weedySeaDragon committed:
refactor TaskAttempt: add Successful..., Failed...

weedySeaDragon committed:
locale entries for refactor TaskAttempt: add Successful..., Failed...

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
ensure shf tasks are defined before running, warn if they are not

weedySeaDragon committed:
ensure there are no TaskAttempts in the db between runs

weedySeaDragon committed:
EvaluatedRakeTask and File: addt'l info about Rake file and tasks

weedySeaDragon committed:
isolate usage of ActivityLog tags

weedySeaDragon committed:
create a Log or one that does nothing (for logging: false)

weedySeaDragon committed:
mock for Rake::Task

weedySeaDragon committed:
spec: create rake files; create a set of known rake files and tasks

weedySeaDragon committed:
isolate behavior that changes an EvaluatedTask (adds duplicate | prev. run)

weedySeaDragon committed:
TasksFinder

weedySeaDragon committed:
TasksRunner

weedySeaDragon committed:
delete spec for orig task written to run one-time tasks; use Classes now

weedySeaDragon committed:
task to create some example one-time task .rake files

weedySeaDragon committed:
task to run 'one time' tasks: shf:one_time:run_onetime_tasks

weedySeaDragon committed:
removed unused var assignments; use File.exist?

weedySeaDragon committed:
hound suggestions to remove cruft and improve

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
add/update comments for instance vars

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
log file name: if self is an instance, use class name

weedySeaDragon committed:
delete log files created during tests

patmbolger commented on app/models/one_time_tasker/task_attempt.rb:
```diff
@@ -0,0 +1,65 @@
+      type_is_valid       = self.was_successful.is_a?(TrueClass) || self.was_successful.is_a?(FalseClass)
+
+      errors.add(:was_successful, I18n.t(i18n_error_key, value: self.was_successful)) unless type_is_valid
+      has_been_attempted = !self.attempted_on.nil?
```
Would `self.attempted_on.present?` read easier than `!self.attempted_on.nil?` ?

patmbolger commented on config/deploy.rb:
```diff
@@ -49,38 +55,46 @@
-      end
-    end
 
+  desc 'run any one-time tasks for this year and this quaters'
```
spelling ...

patmbolger commented on config/initializers/one_time_tasks_runner.rb:
```diff
@@ -0,0 +1,27 @@
+    # from being run again.
+    #
+    # Default is '.ran'
+    config.successful_rakefile_extension = '.ran'
```
Should this be configurable or a fixed constant?

If the former, would changing this at some point in the future would cause the previously-run rake files to be attempted again?  (or, is the mere fact that the filename for each does _not_ end in `.rake` prevent that?)

weedySeaDragon replied:
Only files that end in `.rake` will be attempted (read, in fact).  So any ending that is not `.rake` can be anything.

A fixed-constant makes sense for a default value, here.  The end user(s) should be able to specify what string to use, so that's a configuration variable.

weedySeaDragon replied:
Just FYI: here's the method that gets the list of `.rake` files to evaluate:
`OneTimeTasker::TasksFinder one_time_rake_files`
https://github.com/AgileVentures/shf-project/pull/686/files#diff-ffa7a456562d2fad40916a1d3a0b5f06R264

patmbolger commented on db/migrate/20190601004310_create_one_time_tasker_task_attempts.rb:
```diff
@@ -0,0 +1,13 @@
+      t.timestamps
+    end
+  end
+end
```
Relative to my previous comments, I'm still uneasy about using the DB as our source of knowledge as to which tasks have been run and which have not.  The assumption here is that the DB accurately specifies which tasks have been run and which have not, on _this instance_ of the DB.

For production we're probably OK with that assumption (although we'd be in trouble if - for some reason - we failed to recover this table when restoring the DB after some disaster - but that's a low probability event).

The problem exists, however, for development and staging environments.  If I clone the project, run seeds and then run this, I could get all kinds of strange errors - at rake runtime, and maybe later with data that was altered by a one-time task that should not have been.   Or, I might run the task later when I've build up a lot of local data, and that data might be altered in ways not intended.

I am wondering if we should restrict this rake task to running _only in production_?

weedySeaDragon replied:
In this app, in any environment, if you run a rake task that puts the data into a bad state, then your data is in a bad state.  
Protecting against putting data in a bad state is a _big_ task and not something we've ever really discussed or devoted time to.  (There are definitely apps where that is _critical_ and fundamental, but I don't think this is one of them.) 
For sure, `db:seed` is something that really should only ever been run once in production.  We have guards in there to not run in production really as a brute-force (and common) way to ensure that it is not run again. We use that as a simple, brute-force way to implement "run once."

So I think that is outside of the scope of this.  

If you feel strongly, then can we make that a separate PR so that we can contain the scope of this and try to get this PR wrapped up?

(As an aside: The purpose of a staging environment is to mirror the production environment as closely as possible so that any errors are caught in staging; they never get past staging to get into production.  (Plus, we don't have a staging environment.)  Maybe you mean to refer to "test"?)

weedySeaDragon replied:
Re: the DB

Is your concern about using the DB more about using the application's db to keep track of the state of the data in the application?  IOW, is this more about keeping the information about the "meta state" of the data  outside of the application?

That is definitely something to consider.

The value of using a DB is that you get all of the capabilities and power of a DB (obv): the timestamps, the capabilities of db-management (dumps, schema control, etc etc)., ability to put the DB in separate volumes/disks/etc.

Alternatives would be to track things in a text file (e.g. YAML).  

Worth discussing, but would like to control the scope of this PR, so perhaps this could be done as a next step.

patmbolger replied:
My concern is using the DB for keeping track of the operations (rake tasks) that have been run against the data in that DB.  If the meta data is incomplete or missing, and a user runs the master run-once rake task, that inadvertent changes could occur to the DB.

For production, we're probably OK - since we control that DB and the running of rake tasks.

In development, we could run into the situation where the user has seeded the DB and also added much data during development, then runs the master rake task - which might change the dev DB in ways that are not visible but that create problems for the application and test logic.  That's why I think it might be good to guard against this rake tasks from running _outside_ of production.  Any specific run-once rake tasks that the user needs to run can be run individually, and with the user (presumably) aware of why the task should be run and what it will do to the dev DB.

patmbolger reviewed (commented):
Pls see comments.

weedySeaDragon committed:
fix typo

patmbolger reviewed (approved):
I agree that any other changes we might want to make can be done in a separate effort - this PR is good to go otherwise.

weedySeaDragon commented:
Thanks @patmbolger  
I've added a task in PT to discuss and decide on these issues. https://www.pivotaltracker.com/story/show/167769904
If I missed anything in the PT description, go ahead and edit.  I've added you as an owner as you'll no doubt see. :-)

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
comments; clarify example tasks created (names)

weedySeaDragon committed:
set/create the log first so it can be passed to the TaskStateUpdater

weedySeaDragon committed:
add task to deal with existing (already run) one-time tasks; refactor shared code

weedySeaDragon committed:
deploy: change the desc. for the run one-time task (not based on date anymore)

weedySeaDragon committed:
ensure set_prev_onetime_tasks_as_ran only runs once: write a YAML file

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

# Conflicts:
#	db/structure.sql

weedySeaDragon committed:
Merge remote-tracking branch 'AgileVentures/develop' into sprint68-cap-deploy-1time-tasks-#166136006

# Conflicts:
#	config/locales/en.yml
#	db/structure.sql

weedySeaDragon committed:
rename method to clarify/shorten

weedySeaDragon committed:
doc: flowchart for setting existing one-time tasks to already ran

weedySeaDragon commented:
@patmbolger  This is the last bit needed for this:  We need to incorporate all of the one time tasks that have already been run.  We need to make entries for them in the db (as being already successfully run) and rename their `.rake` file to `.rake.ran` so that they will not be run again.

I created a task that does this:  `shf:one_time:set_prev_onetime_tasks_as_ran` Once this runs successfully it will create a YAML file with this task name and the time it ran.  
The existence of this YAML file is how we know if this task has been run or not.  

I chose a YAML file and a simple structure for showing if the task has been run because if we decide to record info about running one-time tasks outside of the SHF database (e.g. in a future PR), this is an initial stab at how we might record the info.  (And if we do, we'll need to change this logic to look in the contents of the YAML file for this specific info, not just the existence of it.)

I created a simple flow chart to show the logic of this.  I wanted to try to save you a little review time and also document this a bit more. (his file is part of this PR as `docs/dev/OneTimeTasker-flow-for-setting-tasks-to-already-ran.png` so we can add it to the wiki if needed.)

![OneTimeTasker-flow-for-setting-tasks-to-already-ran](https://user-images.githubusercontent.com/673794/63810700-39d1fe80-c8da-11e9-8ced-1b6eb45b3797.png)

weedySeaDragon committed:
doc: fix flowchart and the filename

patmbolger reviewed (approved):
Looks good!
